File: app/src/main/java/com/example/nasmovie/ui/SearchFragment.java
package com.example.nasmovie.ui;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.cardview.widget.CardView;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.example.nasmovie.R;
import com.example.nasmovie.data.local.SearchHistoryManager;
import com.example.nasmovie.data.model.Movie;
import com.example.nasmovie.data.repository.MovieRepository;
import com.example.nasmovie.ui.adapter.SearchResultAdapter;
import com.example.nasmovie.view.NasToolbar;
import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * 搜索 Fragment
 */
public class SearchFragment extends Fragment implements
        SearchResultAdapter.OnMovieClickListener {

    private EditText editSearch;
    private ImageView btnClear;
    private ChipGroup chipGroupHistory;
    private RecyclerView recyclerViewResults;
    private SwipeRefreshLayout swipeRefreshLayout;
    private ProgressBar progressBar;
    private View emptyView;
    private View searchHistoryContainer;
    private NasToolbar toolbar;

    private SearchResultAdapter resultAdapter;
    private MovieRepository repository;
    private SearchHistoryManager historyManager;
    private List<Movie> allMovies = new ArrayList<>();
    private List<Movie> searchResults = new ArrayList<>();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_search, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        initViews(view);
        initData();
        setupSearchHistory();
    }

    private void initViews(View view) {
        toolbar = view.findViewById(R.id.toolbar);
        editSearch = view.findViewById(R.id.edit_search);
        btnClear = view.findViewById(R.id.btn_clear);
        chipGroupHistory = view.findViewById(R.id.chip_group_history);
        recyclerViewResults = view.findViewById(R.id.recycler_view_results);
        swipeRefreshLayout = view.findViewById(R.id.swipe_refresh_layout);
        progressBar = view.findViewById(R.id.progress_bar);
        emptyView = view.findViewById(R.id.empty_view);
        searchHistoryContainer = view.findViewById(R.id.search_history_container);

        // 设置 Toolbar
        AppCompatActivity activity = (AppCompatActivity) getActivity();
        if (activity != null) {
            activity.setSupportActionBar(toolbar.getToolbar());
            if (activity.getSupportActionBar() != null) {
                activity.getSupportActionBar().setDisplayShowTitleEnabled(false);
            }
        }

        // 设置标题和返回按钮
        toolbar.setTitle("搜索");
        toolbar.setShowBack(true);
        toolbar.setOnBackClickListener(() -> {
            if (getActivity() instanceof MainActivity) {
                ((MainActivity) getActivity()).performRealBack();
            }
        });

        // 搜索结果网格 - 3 列
        resultAdapter = new SearchResultAdapter();
        resultAdapter.setOnMovieClickListener(this);
        GridLayoutManager gridLayoutManager = new GridLayoutManager(getContext(), 3);
        recyclerViewResults.setLayoutManager(gridLayoutManager);
        recyclerViewResults.setAdapter(resultAdapter);

        // 搜索输入监听
        editSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                btnClear.setVisibility(s.length() > 0 ? View.VISIBLE : View.GONE);
            }

            @Override
            public void afterTextChanged(Editable s) {}
        });

        // 搜索按钮监听
        editSearch.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                performSearch(editSearch.getText().toString().trim());
                return true;
            }
            return false;
        });

        // 清除按钮
        btnClear.setOnClickListener(v -> {
            editSearch.setText("");
            showSearchHistory();
        });

        // 清空历史按钮
        view.findViewById(R.id.btn_clear_history).setOnClickListener(v -> {
            historyManager.clearHistory();
            setupSearchHistory();
        });

        // 下拉刷新
        swipeRefreshLayout.setOnRefreshListener(() -> {
            String query = editSearch.getText().toString().trim();
            if (!query.isEmpty()) {
                performSearch(query);
            } else {
                swipeRefreshLayout.setRefreshing(false);
            }
        });
        swipeRefreshLayout.setColorSchemeResources(R.color.colorPrimary);
    }

    private void initData() {
        repository = new MovieRepository(requireContext());
        historyManager = new SearchHistoryManager(requireContext());
        loadAllMovies();
    }

    private void loadAllMovies() {
        new Thread(() -> {
            allMovies = repository.getAllMovies();
            requireActivity().runOnUiThread(() -> showSearchHistory());
        }).start();
    }

    private void setupSearchHistory() {
        chipGroupHistory.removeAllViews();
        List<String> searchHistory = historyManager.getSearchHistory();

        for (String history : searchHistory) {
            Chip chip = new Chip(requireContext());
            chip.setText(history);
            chip.setChipBackgroundColorResource(R.color.colorSurface);
            chip.setChipStrokeWidth(1f);
            chip.setChipStrokeColorResource(R.color.divider);
            chip.setTextColor(requireContext().getColor(R.color.textPrimary));
            chip.setOnClickListener(v -> {
                editSearch.setText(history);
                performSearch(history);
            });
            chipGroupHistory.addView(chip);
        }
    }

    private void performSearch(String query) {
        if (query.isEmpty()) {
            showSearchHistory();
            return;
        }

        progressBar.setVisibility(View.VISIBLE);
        searchHistoryContainer.setVisibility(View.GONE);
        swipeRefreshLayout.setRefreshing(false);

        new Thread(() -> {
            // 保存搜索记录
            historyManager.addSearchRecord(query);
            searchResults = repository.searchMovies(query);
            requireActivity().runOnUiThread(() -> {
                progressBar.setVisibility(View.GONE);
                resultAdapter.setMovies(searchResults);

                // 刷新搜索历史显示（包含刚添加的记录）
                setupSearchHistory();

                if (searchResults.isEmpty()) {
                    showEmptyState(query);
                } else {
                    showResults();
                }
            });
        }).start();
    }

    private void showSearchHistory() {
        recyclerViewResults.setVisibility(View.GONE);
        emptyView.setVisibility(View.GONE);
        searchHistoryContainer.setVisibility(View.VISIBLE);
    }

    private void showResults() {
        recyclerViewResults.setVisibility(View.VISIBLE);
        emptyView.setVisibility(View.GONE);
        searchHistoryContainer.setVisibility(View.GONE);
    }

    private void showEmptyState(String query) {
        recyclerViewResults.setVisibility(View.GONE);
        emptyView.setVisibility(View.VISIBLE);
        searchHistoryContainer.setVisibility(View.GONE);

        TextView tvEmptyTitle = emptyView.findViewById(R.id.tv_empty_title);
        TextView tvEmptyDesc = emptyView.findViewById(R.id.tv_empty_desc);

        tvEmptyTitle.setText(getString(R.string.search_no_result));
        tvEmptyDesc.setText("未找到包含 \"" + query + "\" 的电影");
    }

    @Override
    public void onMovieClick(Movie movie) {
        if (getActivity() instanceof MainActivity) {
            ((MainActivity) getActivity()).openDetail(movie.getId());
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        // 刷新搜索历史
        setupSearchHistory();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (repository != null) {
            repository.close();
        }
    }
}

File: app/src/main/java/com/example/nasmovie/ui/SearchFragment.kt
package com.example.nasmovie.ui

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.GridLayoutManager
import com.example.nasmovie.R
import com.example.nasmovie.data.model.Movie
import com.example.nasmovie.databinding.FragmentSearchBinding
import com.example.nasmovie.ui.adapter.SearchResultAdapter
import com.google.android.material.chip.Chip

/**
 * 搜索 Fragment
 */
class SearchFragment : Fragment(), SearchResultAdapter.OnMovieClickListener {

    private var _binding: FragmentSearchBinding? = null
    private val binding get() = _binding!!

    private lateinit var resultAdapter: SearchResultAdapter
    private lateinit var viewModel: SearchViewModel

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentSearchBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        viewModel = ViewModelProvider(this)[SearchViewModel::class.java]
        initViews(view)
        observeViewModel()
    }

    private fun initViews(view: View) {
        // 设置 Toolbar
        val activity = activity as? AppCompatActivity
        activity?.let {
            it.setSupportActionBar(binding.toolbar.toolbar)
            it.supportActionBar?.setDisplayShowTitleEnabled(false)
        }

        // 设置标题和返回按钮
        binding.toolbar.setTitle("搜索")
        binding.toolbar.setShowBack(true)
        binding.toolbar.setOnBackClickListener {
            (activity as? MainActivity)?.performRealBack()
        }

        // 搜索结果网格 - 3 列
        resultAdapter = SearchResultAdapter()
        resultAdapter.setOnMovieClickListener(this)
        val gridLayoutManager = GridLayoutManager(context, 3)
        binding.recyclerViewResults.layoutManager = gridLayoutManager
        binding.recyclerViewResults.adapter = resultAdapter

        // 搜索输入监听
        binding.editSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                binding.btnClear.visibility = if (s?.isNotEmpty() == true) View.VISIBLE else View.GONE
            }
            override fun afterTextChanged(s: Editable?) {}
        })

        // 搜索按钮监听
        binding.editSearch.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                viewModel.performSearch(binding.editSearch.text.toString().trim())
                true
            } else {
                false
            }
        }

        // 清除按钮
        binding.btnClear.setOnClickListener {
            binding.editSearch.setText("")
            viewModel.clearSearchResults()
        }

        // 清空历史按钮
        view.findViewById<View>(R.id.btn_clear_history).setOnClickListener {
            viewModel.clearHistory()
        }

        // 下拉刷新
        binding.swipeRefreshLayout.setOnRefreshListener {
            val query = binding.editSearch.text.toString().trim()
            if (query.isNotEmpty()) {
                viewModel.performSearch(query)
            } else {
                binding.swipeRefreshLayout.isRefreshing = false
            }
        }
        binding.swipeRefreshLayout.setColorSchemeResources(R.color.colorPrimary)
    }

    private fun observeViewModel() {
        viewModel.isLoading.observe(viewLifecycleOwner) { isLoading ->
            if (isLoading) {
                binding.progressBar.visibility = View.VISIBLE
                binding.searchHistoryContainer.visibility = View.GONE
                binding.swipeRefreshLayout.isRefreshing = false
            } else {
                binding.progressBar.visibility = View.GONE
            }
        }

        viewModel.searchHistory.observe(viewLifecycleOwner) { history ->
            setupSearchHistory(history)
        }

        viewModel.searchResults.observe(viewLifecycleOwner) { results ->
            if (results == null) {
                showSearchHistory()
            } else {
                resultAdapter.setMovies(results)
                if (results.isEmpty()) {
                    showEmptyState(binding.editSearch.text.toString().trim())
                } else {
                    showResults()
                }
            }
        }
    }

    private fun setupSearchHistory(searchHistory: List<String>?) {
        binding.chipGroupHistory.removeAllViews()

        searchHistory?.forEach { history ->
            val chip = Chip(requireContext()).apply {
                text = history
                setChipBackgroundColorResource(R.color.colorSurface)
                chipStrokeWidth = 1f
                setChipStrokeColorResource(R.color.divider)
                setTextColor(requireContext().getColor(R.color.textPrimary))
                setOnClickListener {
                    binding.editSearch.setText(history)
                    viewModel.performSearch(history)
                }
            }
            binding.chipGroupHistory.addView(chip)
        }
    }

    private fun showSearchHistory() {
        binding.recyclerViewResults.visibility = View.GONE
        binding.emptyView.visibility = View.GONE
        binding.searchHistoryContainer.visibility = View.VISIBLE
    }

    private fun showResults() {
        binding.recyclerViewResults.visibility = View.VISIBLE
        binding.emptyView.visibility = View.GONE
        binding.searchHistoryContainer.visibility = View.GONE
    }

    private fun showEmptyState(query: String) {
        binding.recyclerViewResults.visibility = View.GONE
        binding.emptyView.visibility = View.VISIBLE
        binding.searchHistoryContainer.visibility = View.GONE

        val tvEmptyTitle = binding.emptyView.findViewById<android.widget.TextView>(R.id.tv_empty_title)
        val tvEmptyDesc = binding.emptyView.findViewById<android.widget.TextView>(R.id.tv_empty_desc)

        tvEmptyTitle.text = getString(R.string.search_no_result)
        tvEmptyDesc.text = "未找到包含 \"$query\" 的电影"
    }

    override fun onMovieClick(movie: Movie) {
        (activity as? MainActivity)?.openDetail(movie.id)
    }

    override fun onResume() {
        super.onResume()
        // 刷新搜索历史
        viewModel.loadSearchHistory()
    }
}